tools/autograd/gen_python_functions.py

```diff
@@ -115,230 +107,218 @@ def should_generate_py_binding(f: NativeFunction) -> bool:
-    return 'namespace' in declaration['method_of']
-
-
-def is_nn_module_function(declaration):
-    return declaration.get('python_module') == 'nn'
-
-
-def is_fft_module_function(declaration):
-    return declaration.get('python_module') == 'fft'
-
+    })
 
-def is_linalg_module_function(declaration):
-    return declaration.get('python_module') == 'linalg'
-
-
-def op_name(declaration):
-    name = declaration['name']
-    if has_outputs(declaration):
-        if not name.endswith("_out"):
-            raise RuntimeError(
-                '{} has output params, expecting name ending with \'_out\''.
-                format(declaration['name']))
-        return name[:-4]
-    else:
-        if name.endswith("_out"):
-            raise RuntimeError(
-                '{}: name ends with \'_out\', expecting output params'.
-                format(declaration['name']))
-        return name
-
-
-def is_py_variable_method(f: NativeFunction) -> bool:
-    return f.python_module is None and Variant.method in f.variants
-
-def is_py_torch_function(f: NativeFunction) -> bool:
-    return f.python_module is None and Variant.function in f.variants
+def load_signatures(
+    native_yaml_path: str,
+    deprecated_yaml_path: str,
+    *,
+    method: bool,
+) -> Sequence[PythonSignatureNativeFunctionPair]:
+    native_functions = list(filter(should_generate_py_binding, parse_native_yaml(native_yaml_path)))
 
-def is_py_nn_function(f: NativeFunction) -> bool:
-    return f.python_module == 'nn'
+    @with_native_function
+    def gen_signature_pairs(f: NativeFunction) -> PythonSignatureNativeFunctionPair:
+        return PythonSignatureNativeFunctionPair(
+            signature=signature(f, method=method),
+            function=f,
+        )
 
-def is_py_fft_function(f: NativeFunction) -> bool:
-    return f.python_module == 'fft'
+    pairs = list(map(gen_signature_pairs, native_functions))
+    deprecated = load_deprecated_signatures(pairs, deprecated_yaml_path, method=method)
+    return pairs + deprecated
 
-def is_py_linalg_function(f: NativeFunction) -> bool:
-    return f.python_module == 'linalg'
+def load_deprecated_signatures(
+    pairs: Sequence[PythonSignatureNativeFunctionPair],
+    deprecated_yaml_path: str,
+    *,
+    method: bool,
+) -> List[PythonSignatureNativeFunctionPair]:
+    # The deprecated.yaml doesn't have complete type information, we need
+    # find and leverage the original ATen signature (to which it delegates
+    # the call) to generate the full python signature.
+    # We join the deprecated and the original signatures using type-only form.
+
+    # native function -> type-only signature
+    @with_native_function
+    def signature_original(f: NativeFunction) -> str:
+        # remove inplace suffix but keep outplace suffix
+        opname = str(f.func.name.name.base)
+        if f.func.is_out_fn():
+            opname += '_out'
+        args = CppSignatureGroup.from_schema(f.func, method=False).signature.arguments()
```

Comment:
@bhosmer updated this according to our offline discussion - it turns out that a.argument could be TensorOptionsArguments (and I cannot simply access `a.argument.type` because of it).